### ISSUE TITLE
Array.prototype.find returns a possibly undefined value even if type guard ensures that the array has one or more satisfying elements.

### ISSUE THREAD

**Issue by yudai-nkt** [microsoft/TypeScript#47404]
# Bug Report

### 🔎 Search Terms

array, some, find, filter, type, guard, predicate

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Type System Behavior and Type Guards.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.4&noUncheckedIndexedAccess=true#code/C4TwDgpgBA8g1lAvFA3gKClA9nAXFYAJwFcIBuDKANwEMAbfYgOziawHcmKBfC0SKAFFChJKko58AM3oBncpQgj8wwlkI8KaACYQAxnRqFoerE1nAoAS1nx8ACiMBzRizacAlEgB8UZ9dlYOApdAyMTMwsoY1liOmBZB3goAB8hEQ8AbQBdLSspKHsYuISAOlksAFsIext4Dy90TFNzSxRqeihuMWL42VKpKyZtWts4Dx40IA)

### 💻 Code

```ts
type Ok = {
  ok: true;
  val: unknown;
};
type Err = {
  ok: false;
  err: Error;
};

declare const isOk: (arg: unknown) => arg is Ok;
declare const results: (Ok | Err)[];

if (results.some(isOk)) {
  const { val } = results.find(isOk);
  //      ~~~
  // error TS2339: Property 'val' does not exist on type 'Ok | undefined'.
}
```

### 🙁 Actual behavior

An error `TS2339: Property 'val' does not exist on type 'Ok | undefined'.` is raised. Since `isOk` returns a type predicate, `results`  inside the `if` statement is guaranteed to have at least one `Ok`. So the destructuring assignment should work.

### 🙂 Expected behavior

No compile errors. The reason is explained above.

### Additional notes

- If `compilerOptions.noUncheckedIndexedAccess` is set to `true`, the same problem happens with `const [{ val }] = results.filter(isOk);` as well.
- The closest issue I found is #46894, but I can't tell with confidence if this is really a duplicate or just looks similar. Feel free to close if the former is the case.

**Comment by MartinJohns** [comment]
IMO it's a duplicate of #46894. That issue is about `some` narrowing the argument, same as yours.

**Comment by yudai-nkt** [comment]
Thanks for the response. I'll close this and keep track of #46894.
